3.2/root/usr/share/container-scripts/mongodb/init-petset-replset.sh

```diff
@@ -90,83 +74,60 @@ function setup_primary() {
-    replset="${MONGODB_REPLICA_NAME}/$(find_endpoints | paste -s -d,)"
-
-    info "Determining primary node ..."
-
-    local primary_host=''
-    local max_attempts=20
-    local attempt
-    for attempt in $(seq 1 ${max_attempts}); do
-        info "Connecting to replica set (attempt #${attempt} from ${max_attempts}) ..."
-        primary_host="$(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host "${replset}" --eval "${FIND_PRIMARY_HOST}" --quiet || :)"
-
-        if [ -n "${primary_host}" ]; then
-            # Unfortunately mongo client (at least in 2.4 version) outputs a lot of
-            # unrelated log messages and here we're filtering them out by
-            # finding line that starts with "PRIMARY|". We also remove port
-            # number at the end.
-            primary_host="$(echo "${primary_host}" | grep '^PRIMARY|' || :)"
-            if [ -n "${primary_host}" ]; then
-                primary_host="$(echo "${primary_host}" | cut -d'|' -f2 | cut -d':' -f1)"
-                info "Found primary node: ${primary_host}"
-                break
-            fi
-        fi
-
-        if [ ${attempt} -lt ${max_attempts} ]; then
-            info "Connection failed, waiting ${SLEEP_TIME} sec before trying again ..."
-            sleep ${SLEEP_TIME}
-            continue
-        fi
-
-        info "ERROR: couldn't determine primary node!"
-        return 1
-    done
-
-    info "Adding ${member_host} to replica set ..."
-
-    # Unfortunately we can't rely on exit code and have to inspect the output
-    local rs_add_out
-    rs_add_out="$(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host "${primary_host}" --eval "${ADD_MEMBER}" --quiet || :)"
-
-    if ! echo "${rs_add_out}" | fgrep -xqs '{ "ok" : 1 }'; then
-        info "ERROR: couldn't join to replica set!"
-        info "DETAILS: command output was:"
-        echo "${rs_add_out}"
-        return 1
-    fi
-
-    info "Successfully started process of joining to replica set"
-    info "Waiting for becoming a secondary node ..."
-
-    local rs_status_out
-    rs_status_out="$(mongo admin -u admin -p "${MONGODB_ADMIN_PASSWORD}" --host "${primary_host}" --eval "${WAIT_SECONDARY_STATUS}" --quiet || :)"
-
-    if ! echo "${rs_status_out}" | fgrep -xqs 'SECONDARY'; then
-        info "ERROR: couldn't join to replica set!"
-        info "CAUSE: failed after waiting for becoming a secondary node. Command output was:"
-        echo "${rs_status_out}"
-        return 1
-    fi
-
-    info "Successfully joined replica set"
+function add_member() {
+  local host="$1"
+  info "Adding ${host} to replica set ..."
+
+  local script
+  script="
+    var ret = rs.add('${host}');
+    if (ret.ok) {
+      quit(0);
+    }
+    // ignore error if host is already in the configuration
+    if (ret.code == 163) {
+      quit(0);
+    }
+    printjson(ret);
+    quit(1);
+  "
+
+  # TODO: replace this with a call to `replset_addr` from common.sh, once it returns host names.
```

Comment:
> But IMPOV it does not matter while connecting replset.

That's a good point. :+1: 

It might work without changing `replset_addr`. I'll try.

Thanks!